Update StringView::getCharacters() to take in a span

cdumez opened the pull request:
#### 7a039c8cf8011ef0a2a405b7a3ada7dbd0db6f43
<pre>
Update StringView::getCharacters() to take in a span
<a href="https://bugs.webkit.org/show_bug.cgi?id=283099">https://bugs.webkit.org/show_bug.cgi?id=283099</a>

Reviewed by Geoffrey Garen.

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::characters):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::setKeywordValue):
* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSRopeString::resolveRopeInternalNoSubstring const):
(JSC::JSRopeString::resolveRopeToAtomString const):
(JSC::JSRopeString::resolveRopeToExistingAtomString const):
(JSC::JSRopeString::resolveRopeWithFunction const):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::repeatCharacter):
(JSC::JSRopeString::resolveToBufferSlow):
(JSC::JSRopeString::resolveToBuffer):
(JSC::jsAtomString):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData):
(JSC::appendStringToDataWithOneCharacterSeparatorRepeatedly):
(JSC::joinStrings):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
(JSC::normalize):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::jsSpliceSubstringsWithSeparators):
* Source/WTF/wtf/HexNumber.h:
(WTF::StringTypeAdapter&lt;HexNumberBuffer&gt;::writeTo const):
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter&lt;uint64_t&gt;::writeTo const):
* Source/WTF/wtf/UUID.h:
(WTF::StringTypeAdapter&lt;UUID&gt;::writeTo const):
* Source/WTF/wtf/text/Base64.h:
(WTF::StringTypeAdapter&lt;Base64Specification&gt;::writeTo const):
* Source/WTF/wtf/text/MakeString.h:
(WTF::tryMakeStringImplFromAdaptersInternal):
(WTF::tryMakeAtomStringFromAdapters):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::extendBufferForAppendingLChar):
(WTF::StringBuilder::extendBufferForAppendingWithUpconvert):
(WTF::StringBuilder::append):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::appendFromAdapters):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::allocateBuffer):
(WTF::StringBuilder::extendBufferForAppending):
(WTF::StringBuilder::extendBufferForAppendingSlowCase):
* Source/WTF/wtf/text/StringBuilderJSON.cpp:
(WTF::StringBuilder::appendQuotedJSONString):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::StringTypeAdapter&lt;std::tuple&lt;StringTypes::writeTo const):
(WTF::StringTypeAdapter&lt;PaddingSpecification&lt;UnderlyingElementType&gt;&gt;::writeTo const):
(WTF::stringTypeAdapterAccumulator):
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter&lt;FormattedNumber&gt;::writeTo const):
(WTF::StringTypeAdapter&lt;FormattedCSSNumber&gt;::writeTo const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryCreateUninitialized):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::getCharacters8 const):
(WTF::StringView::getCharacters16 const):
(WTF::StringView::getCharacters const):
(WTF::append):
* Source/WTF/wtf/text/cf/StringConcatenateCF.h:
(WTF::StringTypeAdapter&lt;CFStringRef&gt;::writeTo&lt;LChar&gt; const):
(WTF::StringTypeAdapter&lt;CFStringRef&gt;::writeTo&lt;UChar&gt; const):
* Source/WTF/wtf/text/cocoa/ContextualizedNSString.mm:
(-[WTFContextualizedNSString getCharacters:range:]):
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::setSubprotocol):
(WebCore::ThreadableWebSocketChannelClientWrapper::setExtensions):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::normalizeSpaces):
* Source/WebCore/platform/ProcessQualified.h:
(WTF::ProcessQualifiedStringTypeAdapter::writeTo const):
* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::centerTruncateToBuffer):
(WebCore::rightTruncateToBuffer):
(WebCore::rightClipToCharacterBuffer):
(WebCore::rightClipToWordBuffer):
(WebCore::leftTruncateToBuffer):
(WebCore::truncateString):
* Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
(WebCore::FontCache::systemFallbackForCharacterCluster):
(WebCore::createGDIFont):
(WebCore::FontCache::getFontSelectionCapabilitiesInFamily):
* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
(WebCore::createGlobalData):
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::fileSystemPathFromURLOrTitle):
(WebCore::Pasteboard::writeURLToDataObject):
(WebCore::createGlobalImageFileDescriptor):
* Source/WebGPU/WGSL/Types.h:
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetCharacters):
* Tools/TestWebKitAPI/Tests/WTF/MoveOnly.h:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp:
(WTF::StringTypeAdapter&lt;WKStringRef&gt;::writeTo&lt;LChar&gt; const):
(WTF::StringTypeAdapter&lt;WKStringRef&gt;::writeTo&lt;UChar&gt; const):

Canonical link: <a href="https://commits.webkit.org/286694@main">https://commits.webkit.org/286694@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47443acb61b2a32ae66ae68bfc15f8e8e3795c44

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76788 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55823 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29694 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81321 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28064 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78905 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64965 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4116 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60173 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18261 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79855 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50119 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65925 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40492 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47521 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23419 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26388 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69971 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68638 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23749 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82765 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76065 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4164 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2757 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68455 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4317 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65897 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67707 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11684 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9767 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98319 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11885 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4111 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6922 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21505 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4134 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7564 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5892 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
```diff
@@ -608,7 +608,7 @@ Vector<FontSelectionCapabilities> FontCache::getFontSelectionCapabilitiesInFamil
     logFont.lfCharSet = DEFAULT_CHARSET;
     StringView truncatedFamily = StringView(familyName).left(static_cast<unsigned>(LF_FACESIZE - 1));
-    truncatedFamily.getCharacters(ucharFrom(logFont.lfFaceName));
+    truncatedFamily.getCharacters(unsafeMakeSpan(ucharFrom(logFont.lfFaceName), truncatedFamily.length()));
```
This unsafeMakeSpan length is kind of a tautology. Would be nicer to have a span() helper that took a LOGFONT as an argument. I guess it's not worth it in this patch. But I'm trying to convey how I think of "good" use of unsafeMakeSpan vs "bad".

cdumez replied:
Yes, I was a bit lazy with the Windows code :) I have now made fixes to use the actual buffer size.

geoffreygaren commented on Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
```diff
@@ -167,7 +167,7 @@ HGLOBAL createGlobalData(const String& str)
     HGLOBAL vm = ::GlobalAlloc(GPTR, (str.length() + 1) * sizeof(UChar));
     if (!vm)
         return 0;
-    UChar* buffer = static_cast<UChar*>(GlobalLock(vm));
+    auto buffer = unsafeMakeSpan(static_cast<UChar*>(GlobalLock(vm)), str.length() + 1);
```
I guess if we were serious about Windows we would make a GlobalAllocSpan data type. But it's probably not worth it for this patch.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Update StringView::getCharacters() to take in a span
https://bugs.webkit.org/show_bug.cgi?id=283099

Reviewed by Geoffrey Garen.

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::characters):
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::setKeywordValue):
* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSRopeString::resolveRopeInternalNoSubstring const):
(JSC::JSRopeString::resolveRopeToAtomString const):
(JSC::JSRopeString::resolveRopeToExistingAtomString const):
(JSC::JSRopeString::resolveRopeWithFunction const):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::repeatCharacter):
(JSC::JSRopeString::resolveToBufferSlow):
(JSC::JSRopeString::resolveToBuffer):
(JSC::jsAtomString):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData):
(JSC::appendStringToDataWithOneCharacterSeparatorRepeatedly):
(JSC::joinStrings):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
(JSC::normalize):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::jsSpliceSubstringsWithSeparators):
* Source/WTF/wtf/HexNumber.h:
(WTF::StringTypeAdapter<HexNumberBuffer>::writeTo const):
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::ObjectIdentifierGenericBaseStringTypeAdapter<uint64_t>::writeTo const):
* Source/WTF/wtf/UUID.h:
(WTF::StringTypeAdapter<UUID>::writeTo const):
* Source/WTF/wtf/text/Base64.h:
(WTF::StringTypeAdapter<Base64Specification>::writeTo const):
* Source/WTF/wtf/text/MakeString.h:
(WTF::tryMakeStringImplFromAdaptersInternal):
(WTF::tryMakeAtomStringFromAdapters):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::extendBufferForAppendingLChar):
(WTF::StringBuilder::extendBufferForAppendingWithUpconvert):
(WTF::StringBuilder::append):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::appendFromAdapters):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::allocateBuffer):
(WTF::StringBuilder::extendBufferForAppending):
(WTF::StringBuilder::extendBufferForAppendingSlowCase):
* Source/WTF/wtf/text/StringBuilderJSON.cpp:
(WTF::StringBuilder::appendQuotedJSONString):
* Source/WTF/wtf/text/StringConcatenate.h:
(WTF::StringTypeAdapter<std::tuple<StringTypes::writeTo const):
(WTF::StringTypeAdapter<PaddingSpecification<UnderlyingElementType>>::writeTo const):
(WTF::stringTypeAdapterAccumulator):
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter<FormattedNumber>::writeTo const):
(WTF::StringTypeAdapter<FormattedCSSNumber>::writeTo const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryCreateUninitialized):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::getCharacters8 const):
(WTF::StringView::getCharacters16 const):
(WTF::StringView::getCharacters const):
(WTF::append):
* Source/WTF/wtf/text/cf/StringConcatenateCF.h:
(WTF::StringTypeAdapter<CFStringRef>::writeTo<LChar> const):
(WTF::StringTypeAdapter<CFStringRef>::writeTo<UChar> const):
* Source/WTF/wtf/text/cocoa/ContextualizedNSString.mm:
(-[WTFContextualizedNSString getCharacters:range:]):
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::setSubprotocol):
(WebCore::ThreadableWebSocketChannelClientWrapper::setExtensions):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::normalizeSpaces):
* Source/WebCore/platform/ProcessQualified.h:
(WTF::ProcessQualifiedStringTypeAdapter::writeTo const):
* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::centerTruncateToBuffer):
(WebCore::rightTruncateToBuffer):
(WebCore::rightClipToCharacterBuffer):
(WebCore::rightClipToWordBuffer):
(WebCore::leftTruncateToBuffer):
(WebCore::truncateString):
* Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
(WebCore::FontCache::systemFallbackForCharacterCluster):
(WebCore::createGDIFont):
(WebCore::FontCache::getFontSelectionCapabilitiesInFamily):
* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
(WebCore::createGlobalData):
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::fileSystemPathFromURLOrTitle):
(WebCore::Pasteboard::writeURLToDataObject):
(WebCore::createGlobalImageFileDescriptor):
* Source/WebGPU/WGSL/Types.h:
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetCharacters):
* Tools/TestWebKitAPI/Tests/WTF/MoveOnly.h:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp:
(WTF::StringTypeAdapter<WKStringRef>::writeTo<LChar> const):
(WTF::StringTypeAdapter<WKStringRef>::writeTo<UChar> const):

Canonical link: https://commits.webkit.org/286694@main
